ring-core/src/ring/middleware/multipart_params.clj

```diff
@@ -135,41 +137,60 @@
-
-  :fallback-encoding - specifies the character encoding used in parsing if a
-                       part of the request does not specify encoding in its
-                       content type or no part named \"_charset_\" is present.
-                       Has no effect if :encoding is also set.
-
-  :store             - a function that stores a file upload. The function
-                       should expect a map with :filename, :content-type and
-                       :stream keys, and its return value will be used as the
-                       value for the parameter in the multipart parameter map.
-                       The default storage function is the temp-file-store.
-
-  :progress-fn       - a function that gets called during uploads. The
-                       function should expect four parameters: request,
-                       bytes-read, content-length, and item-count."
+  The following options are accepted:
+
+  :encoding
+  - Character encoding to use for multipart parsing. Overrides the encoding
+    specified in the request. If not specified,uses the encoding specified in a
+    part named \"_charset_\", or the content type for each part, or request
+    character encoding if the part has no encoding, or \"UTF-8\" if no request
+    character encoding is set.
+
+  :fallback-encoding
+  - Specifies the character encoding used in parsing if a part of the request
+    does not specify encoding in its content type or no part named \"_charset_\"
+    is present.  Has no effect if :encoding is also set.
+
+  :store
+  - A function that stores a file upload. The function should expect a map with
+    :filename, :content-type and :stream keys, and its return value will be used
+    as the value for the parameter in the multipart parameter map. The default
+    storage function is the temp-file-store.
+
+  :progress-fn
+  - A function that gets called during uploads. The function should expect four
+    parameters: request, bytes-read, content-length, and item-count.
+
+  :invalid-filename-handler
+  - A function that gets called when the file being uploaded has an invalid name.
+    The function should expect two parameters: request and an exception of type
+    InvalidFileNameException. It should return a ring response."
   ([handler]
    (wrap-multipart-params handler {}))
   ([handler options]
-   (fn
-     ([request]
-      (handler (multipart-params-request request options)))
-     ([request respond raise]
-      (handler (multipart-params-request request options) respond raise)))))
+   (let [invalid-filename-handler
+         (:invalid-filename-handler options default-invalid-filename-handler)]
+     (fn ([request]
+          (let [req-or-ex (try
+                            (multipart-params-request request options)
+                            (catch Exception ex ex))]
+            (if (instance? Throwable req-or-ex)
+              (invalid-filename-handler request req-or-ex)
```

Comment:
The term "handler" has specific meaning in Ring, so if we call something `invalid-filename-handler`, then it should be a handler function.

Perhaps something like:

```clojure
(invalid-filename-handler (assoc request ::invalid-filename (.getName req-or-ex)))
```

We probably want to factor out some of this code into a function.

Reply:
I opted to add the entire exception to the request, in case people want to e.g. log the exception itself when overriding the default handler.

How do you want me to refactor this?  It's a bit awkward due to due the different arities.  I tried a few variants but it didn't really result in less / more elegant code.

Reply:
I'd rather not add the full exception, because it ties the multipart middleware to the commons upload library, which eventually we want to get rid of.

Reply:
Regarding the refactor, you can move the code that adds the invalid filename to the request into its own private function.